### PR TITLE
Replace flax usage with equinox

### LINKED ISSUE #582: Replace flax usage with equinox
### What's the new feature?

In `coreax.networks.py` we have the class `ScoreNetwork` which uses `flax.linen.Module`, as-well as the use of `flax.training.train_state.TrainState` in the same file. We could instead use Equinox functionality for the same purpose, and remove the need to use flax at all. Note that Equinox is used elsewhere in the codebase anyway. 

### What value does this add?

Reliance on fewer third party packages.

### Is there an alternative you've considered?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jd61349**
### PR Type

- Refactoring (no functional changes)

### Description

Replace all usage of Flax with Equinox. Fixes #582.

### How Has This Been Tested?

Related tests have been changed to reflect updated code, however what they test has remained the same. All tests pass.

### Does this PR introduce a breaking change?

No.

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have ensured my code is easy to understand, including docstrings and comments where necessary.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have updated CHANGELOG.md, if appropriate.

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
`MLP.__init__` now passes `mode="batch"` to `eqx.nn.BatchNorm`, but `pyproject.toml` still supports `equinox>=0.11.11`. In Equinox 0.11.11 `BatchNorm.__init__` has no `mode` argument, so a lowest-resolution install fails as soon as this benchmark constructs the model. Could we either raise the lower bound to the version that introduced this API or avoid the newer argument, and add a lowest-dependency benchmark smoke test?

**Comment by jd61349** [comment]
> `MLP.__init__` now passes `mode="batch"` to `eqx.nn.BatchNorm`, but `pyproject.toml` still supports `equinox>=0.11.11`. In Equinox 0.11.11 `BatchNorm.__init__` has no `mode` argument, so a lowest-resolution install fails as soon as this benchmark constructs the model. Could we either raise the lower bound to the version that introduced this API or avoid the newer argument, and add a lowest-dependency benchmark smoke test?

Good spot! Thanks

**Review by sylvesterkaczmarek — APPROVED** [review]
Re-reviewed current `15a48daf`. The compatibility issue I raised is resolved: the Equinox lower bound is now `>=0.13.0`, matching the `BatchNorm(..., mode="batch")` API used by the benchmark, and the lowest/highest dependency test matrix is green across supported Python versions. No remaining blocker from me.

File: pyproject.toml [hunk 1]
```diff
@@ -64,7 +63,7 @@ doc = [
 test = [
     {include-group = "benchmark"},
     "beartype>=0.19",
-    "pytest>=9",
+    "pytest >= 9, != 9.1, != 9.1.1",
```
Comment (tm91236) [inline pyproject.toml]: Please fix spacing and add a comment explaining why we need to exclude 9.1.0 and 9.1.1

File: pyproject.toml [hunk 2]
```diff
@@ -30,8 +30,7 @@ dependencies = [
     # NOTE: lower bounds are the lowest supported version when using the 'lowest'
     # resolution mode of uv on the latest supported version of Python - the bounds
     # are only sharp for the latest supported Python version.
-    "equinox>=0.11.11",
-    "flax>=0.10.0",
+    "equinox>=0.13.0",
```
Comment (tm91236) [inline pyproject.toml:33]: This is a breaking change so should be added to the change log.

File: .pylintrc [hunk 3]
```diff
@@ -107,18 +107,13 @@ recursive=no
 # source root.
 source-roots=
 
-# When enabled, pylint would attempt to guess common misconfiguration and emit
-# user-friendly hints instead of false-positive error messages.
-suggestion-mode=yes
-
```
Comment (tm91236) [inline .pylintrc:113]: Why was this changed?

**Review by tm91236 — CHANGES_REQUESTED** [review]
This looks good, but it does introduce some breaking changes without any deprecation warnings. I think we should keep the old network.py module, but mark the ScoreNetwork as deprecated for removal in v1.2.0.

File: .pylintrc (hunk 3 above)
Comment (jd61349) [reply .pylintrc:113]: Version 4.0 of Pylint removed it (https://pylint.readthedocs.io/en/latest/whatsnew/4/4.0/index.html):

"""

The suggestion-mode option was removed, as pylint now always emits user-friendly hints instead of false-positive error messages. You should remove it from your conf if it's defined.

Refs #9962 ([#9962](https://github.com/pylint-dev/pylint/issues/9962))

"""

We currently scope as `"pylint>=4.0.4",`.

File: pyproject.toml (hunk 2 above)
Comment (jd61349) [reply pyproject.toml:33]: Added

File: pyproject.toml (hunk 1 above)
Comment (jd61349) [reply pyproject.toml]: Fixed

**Review by sylvesterkaczmarek — CHANGES_REQUESTED** [review]
Re-reviewed current `45e5a4b5`. Restoring `coreax/networks.py` for the requested deprecation path introduces a new compatibility problem: that module imports `flax`, but this PR still removes `flax` from the package dependencies. A normal installation can therefore no longer import the deprecated module that is being kept specifically to preserve compatibility. The current pre-commit run confirms this with unresolved imports for `flax`, `flax.linen`, and `flax.training` in `coreax/networks.py`. Please retain the `flax` dependency for the deprecation window (or otherwise make the compatibility module self-contained) and add a simple normal-install import regression for `coreax.networks`/`ScoreNetwork`.

File: documentation/source/conf.py [hunk 4]
```diff
@@ -162,6 +166,8 @@
 nitpick_ignore = [
     ("py:class", "flax.core.scope.Scope"),
     ("py:class", "flax.linen.module._Sentinel"),
+    ("py:class", "flax.linen.module.Module"),
```
Comment (jd61349) [inline documentation/source/conf.py:169]: docs were failing in workflow without these

**Comment by jd61349** [comment]
> Re-reviewed current `45e5a4b5`. Restoring `coreax/networks.py` for the requested deprecation path introduces a new compatibility problem: that module imports `flax`, but this PR still removes `flax` from the package dependencies. A normal installation can therefore no longer import the deprecated module that is being kept specifically to preserve compatibility. The current pre-commit run confirms this with unresolved imports for `flax`, `flax.linen`, and `flax.training` in `coreax/networks.py`. Please retain the `flax` dependency for the deprecation window (or otherwise make the compatibility module self-contained) and add a simple normal-install import regression for `coreax.networks`/`ScoreNetwork`.

Agreed, thanks! Re-added the removed dependencies, and missing test file.

**Comment by jd61349** [comment]
A note to whoever bumps to the next version, the following needs to be done wrt this PR:

1. Remove `flax`, `orbax-checkpoint`, and `tensorstore` dependencies.
2. Remove `networks.py` and `test_networks.py`
3. Ensure documentation files and nitpicks related to `flax` and `networks.py` are removed.

File: tests/unit/test_solvers.py
Comment (jd61349) [inline tests/unit/test_solvers.py:1]: Pre-commit was failing due to `pylint: disable-next=too-many-positional-arguments`

File: tests/unit/test_kernels.py
Comment (jd61349) [inline tests/unit/test_kernels.py:1]: Pre-commit was failing due to `pylint: disable-next=too-many-positional-arguments`

File: coreax/kernels/scalar_valued.py
Comment (jd61349) [inline coreax/kernels/scalar_valued.py:1]: Pre-commit was failing due to `pylint: disable-next=too-many-positional-arguments`

**Review by sylvesterkaczmarek — APPROVED** [review]
Rechecked current 9b859ebd. The deprecated coreax.networks compatibility path is importable again in a normal install because flax is restored as a package dependency, and test_networks.py directly imports coreax.networks and exercises ScoreNetwork. The full lowest/highest dependency matrix is green. My compatibility blocker is resolved.

File: pyproject.toml [hunk 5]
```diff
@@ -176,6 +176,8 @@ filterwarnings = [
     "error",
     "ignore::ImportWarning:umap",
     "ignore::DeprecationWarning:jaxopt",
+    "ignore:The 'coreax\\.networks' module is deprecated:DeprecationWarning",
+    "ignore:ScoreNetwork will be removed in version 1\\.2\\.0\\.:DeprecationWarning",
```
Comment (tm91236) [inline pyproject.toml]: Can this be `ignore:DeprecationWarning:coreax`?

**Review by tm91236 — CHANGES_REQUESTED** [review]
Please can we remove the tests for the now deprecated ScoreNetwork, and add a new tests to check the deprecation warning is raised when attempting to instantiate the ScoreNetwork.

**Comment by sylvesterkaczmarek** [comment]
One concrete follow-up on current 7eb669b7: the requested removal of the deprecated ScoreNetwork behavior tests has made the coverage gate fail. CI now reports coreax/networks.py at 54% and overall coverage at 97.27148704%, down from the 97.94% baseline by 0.666%, exceeding the repository tolerance. The functional/dependency matrix is green; this is specifically the coverage policy reacting to the intentionally untested deprecated module. That will need reconciling before merge, for example by whatever coverage treatment the maintainers prefer for code retained only for the deprecation window.

**Review by tm91236 — APPROVED** [review]
Thank you for making these changes. Reduction in code coverage is acceptable given the uncovered code is now deprecated.
